geom/twkb_parser.go

```diff
@@ -0,0 +1,528 @@
+	}
+	if uint64(p.pos)+bytesRemaining > uint64(len(p.twkb)) {
+		return fmt.Errorf("remaining input (%d bytes) smaller than size varint indicates (%d bytes)", len(p.twkb)-p.pos, bytesRemaining)
+	}
+	return nil
+}
+
+func (p *twkbParser) parseBBox() error {
+	// Parse the bounding box, but do not otherwise use it.
+	for d := 0; d < p.dimensions; d++ {
+		minVal, err := p.parseSignedVarint()
+		if err != nil {
+			return fmt.Errorf("BBox min varint malformed: %w", err)
+		}
+		p.bbox = append(p.bbox, minVal)
+
+		deltaVal, err := p.parseSignedVarint()
+		if err != nil {
+			return fmt.Errorf("BBox delta varint malformed: %w", err)
+		}
+		p.bbox = append(p.bbox, deltaVal)
+	}
+	return nil
+}
+
+func (p *twkbParser) parsePoint() (Point, error) {
+	if p.isEmpty {
+		return NewEmptyPoint(p.ctype), nil
+	}
+	return p.nextPoint()
+}
+
+func (p *twkbParser) nextPoint() (Point, error) {
+	var c Coordinates
+	c.Type = p.ctype
+
+	coords, err := p.parsePointArray(1)
+	if err != nil {
+		return Point{}, err
+	}
+
+	c.XY.X = coords[0]
+	c.XY.Y = coords[1]
+	if p.hasZ && p.hasM {
+		c.Z = coords[2]
+		c.M = coords[3]
+	} else if p.hasZ {
+		c.Z = coords[2]
+	} else if p.hasM {
+		c.M = coords[2]
+	}
+
+	return NewPoint(c, p.opts...)
+}
+
+func (p *twkbParser) parseLineString() (LineString, error) {
+	if p.isEmpty {
+		return NewLineString(NewSequence(nil, p.ctype), p.opts...)
+	}
+	return p.nextLineString()
+}
+
+func (p *twkbParser) nextLineString() (LineString, error) {
+	coords, _, err := p.parsePointCountAndArray()
+	if err != nil {
+		return LineString{}, err
+	}
+	return NewLineString(NewSequence(coords, p.ctype), p.opts...)
+}
+
+func (p *twkbParser) parsePolygon() (Polygon, error) {
+	if p.isEmpty {
+		return NewPolygon(nil, p.opts...)
+	}
+	return p.nextPolygon()
+}
+
+func (p *twkbParser) nextPolygon() (Polygon, error) {
+	numRings, err := p.parseUnsignedVarint()
+	if err != nil {
+		return Polygon{}, fmt.Errorf("num rings varint malformed: %w", err)
+	}
+
+	var rings []LineString
+	for r := uint64(0); r < numRings; r++ {
+		coords, numPoints, err := p.parsePointCountAndArray()
+		if err != nil {
+			return Polygon{}, err
+		}
+		if numPoints >= 2 {
+			// The spec says rings may need to be closed:
+			// "rings are assumed to be implicitly closed, so the first
+			//  and last point should not be the same"
+
+			// Note, some implementations can generate TWKB with the rings
+			// already closed. We wish to gracefully parse these cases too.
+			// So check if any of the first and final point's coords differ.
+			finalPointDiffersFromFirst := false
+			for d := 0; d < p.dimensions; d++ {
+				first := coords[d]
+				final := coords[d+(numPoints-1)*p.dimensions]
+				if first != final {
+					finalPointDiffersFromFirst = true
+					break
+				}
+			}
+
+			if finalPointDiffersFromFirst {
+				// Append first point again, to close the ring.
+				for d := 0; d < p.dimensions; d++ {
+					coords = append(coords, coords[d])
```

Comment:
Very minor nit: could do `coords = append(coords, coords[:p.dimensions]...)` instead of the loop.